libcoap/src/oscore.rs

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright © The libcoap-rs Contributors, all rights reserved.
+ * This file is part of the libcoap-rs project, see the README file for
+ * general information on this project and the NOTICE.md and LICENSE files
+ * for information regarding copyright ownership and terms of use.
+ *
+ * oscore.rs - Wrapper for libcoap OSCORE functionality.
+ */
+
+use core::{ffi::c_void, ptr};
+
+use libcoap_sys::{coap_delete_oscore_conf, coap_new_oscore_conf, coap_new_str_const, coap_oscore_conf_t};
+
+use crate::error::OscoreConfigError;
+
+/// Represents an oscore config object which stores the underlying
+/// coap_oscore_conf_t C struct.
+pub struct OscoreConf {
+    raw_conf: *mut coap_oscore_conf_t,
+    pub(crate) initial_recipient: Option<String>,
+    drop: bool,
+}
+
+impl OscoreConf {
+    /// Creates a new OscoreConf.
+    ///
+    /// # Errors
+    /// Will return a [OscoreConfigError] if creating the oscore config fails (most likely due to
+    /// invalid oscore config bytes provided).
+    pub fn new(
+        seq_initial: u64,
+        oscore_conf_bytes: &[u8],
+        save_seq_num_func: extern "C" fn(seq_num: u64, _param: *mut c_void) -> i32,
+    ) -> Result<Self, OscoreConfigError> {
+        // Creates the raw_struct containing the config provided by the caller.
+        // SAFETY: Provided pointer and length point to a valid byte string usable by
+        // coap_new_str_const().
+        let conf = unsafe { coap_new_str_const(oscore_conf_bytes.as_ptr(), oscore_conf_bytes.len()) };
+        if conf.is_null() {
+            return Err(OscoreConfigError::Unknown);
+        }
+
+        // SAFETY:
+        // The parts of the byte string referenced by conf are defensively copied if used
+        // by the newly created oscore_conf.
+        // Conf was just checked for invalidity, whether or not it containes all required fields.
+        // - save_seq_num_func is specifically designed to work as a callback for this
+        //   function.
+        // - save_seq_num_func_param may be a null pointer (save_seq_num_func does
+        //   not use it).
+        let oscore_conf = unsafe { coap_new_oscore_conf(*conf, Some(save_seq_num_func), ptr::null_mut(), seq_initial) };
+        if oscore_conf.is_null() {
+            return Err(OscoreConfigError::Unknown);
+        }
```

Comment:
Does `coap_new_oscore_conf()` also free the provided `conf`? Otherwise, we will need to free it after calling `coap_new_oscore_conf()` to avoid a memory leak.